File: frontend/src/types/armada.ts
export type StatusArmada =
  | "standby"
  | "sedang_bertugas"
  | "pemeliharaan"
  | "menunggu_sparepart"
  | "rusak_ringan"
  | "rusak_berat"
  | "tidak_aktif"
  | "menunggu_approval";

export type ApprovalStatus = "tidak_perlu" | "pending" | "disetujui" | "ditolak";

export interface ArmadaListItem {
  id: number;
  kode_armada: string;
  nama_armada: string | null;
  jenis_kendaraan_id: number;
  lokasi_saat_ini_id: number | null;
  status_armada: StatusArmada;
  no_polisi: string | null;
}

export interface ArmadaPublic extends ArmadaListItem {
  merk: string | null;
  type: string | null;
  tahun: number | null;
  no_mesin: string | null;
  no_rangka: string | null;
  no_bpkb: string | null;
  tanggal_stnk: string | null;
  kapasitas: string | null;
  status_kepemilikan: string | null;
  qr_code_value: string;
  status_approval: ApprovalStatus;
  created_at: string;
}

export interface ArmadaPayload {
  kode_armada: string;
  nama_armada?: string | null;
  jenis_kendaraan_id: number;
  merk?: string | null;
  type?: string | null;
  tahun?: number | null;
  no_polisi?: string | null;
  no_lambung?: string | null;
  no_mesin?: string | null;
  no_rangka?: string | null;
  no_bpkb?: string | null;
  tanggal_stnk?: string | null;
  kapasitas?: string | null;
  status_kepemilikan?: string | null;
  lokasi_saat_ini_id?: number | null;
}

export const STATUS_KRITIS: StatusArmada[] = ["rusak_berat", "tidak_aktif"];

export interface HistoriLokasi {
  id: number;
  armada_id: number;
  lokasi_lama_id: number | null;
  lokasi_baru_id: number;
  tanggal_pindah: string;
  dipindahkan_oleh: number | null;
  keterangan: string | null;
}

export interface HistoriStatus {
  id: number;
  armada_id: number;
  status_lama: StatusArmada | null;
  status_baru: StatusArmada;
  tanggal: string;
  diajukan_oleh: number | null;
  butuh_approval: boolean;
  approval_status: ApprovalStatus;
  disetujui_oleh: number | null;
  tanggal_approval: string | null;
  catatan_approval: string | null;
  keterangan: string | null;
}

export const STATUS_LABEL: Record<StatusArmada, string> = {
  standby: "Standby",
  sedang_bertugas: "Sedang Bertugas",
  pemeliharaan: "Pemeliharaan",
  menunggu_sparepart: "Menunggu Sparepart",
  rusak_ringan: "Rusak Ringan",
  rusak_berat: "Rusak Berat",
  tidak_aktif: "Tidak Aktif",
  menunggu_approval: "Menunggu Approval",
};

File: backend/app/schemas/armada.py
from datetime import date, datetime

from pydantic import BaseModel

from app.models.armada import ApprovalStatus, StatusArmada


class ArmadaBase(BaseModel):
    kode_armada: str
    nama_armada: str | None = None
    jenis_kendaraan_id: int
    merk: str | None = None
    type: str | None = None
    tahun: int | None = None
    no_polisi: str | None = None
    no_lambung: str | None = None
    no_mesin: str | None = None
    no_rangka: str | None = None
    no_bpkb: str | None = None
    tanggal_stnk: date | None = None
    kapasitas: str | None = None
    status_kepemilikan: str | None = None
    lokasi_saat_ini_id: int | None = None
    driver_id: int | None = None


class ArmadaCreate(ArmadaBase):
    pass


class ArmadaUpdate(BaseModel):
    kode_armada: str | None = None
    nama_armada: str | None = None
    jenis_kendaraan_id: int | None = None
    merk: str | None = None
    type: str | None = None
    tahun: int | None = None
    no_polisi: str | None = None
    no_lambung: str | None = None
    no_mesin: str | None = None
    no_rangka: str | None = None
    no_bpkb: str | None = None
    tanggal_stnk: date | None = None
    kapasitas: str | None = None
    status_kepemilikan: str | None = None
    lokasi_saat_ini_id: int | None = None
    driver_id: int | None = None


class ArmadaPublic(ArmadaBase):
    id: int
    qr_code_value: str
    status_armada: StatusArmada
    status_approval: ApprovalStatus
    created_at: datetime
    driver_name: str | None = None
    operator_lapangan_id: int | None = None

    class Config:
        from_attributes = True


class ArmadaListItem(BaseModel):
    id: int
    kode_armada: str
    nama_armada: str | None = None
    jenis_kendaraan_id: int
    lokasi_saat_ini_id: int | None = None
    status_armada: StatusArmada
    no_polisi: str | None = None
    tanggal_stnk: date | None = None
    driver_id: int | None = None
    driver_name: str | None = None

    class Config:
        from_attributes = True
